src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,271 @@
+                        int space = lastLine.indexOf(' ');
+                        lastId = space == -1 ? null : lastLine.substring(space + 1);
+                    }
+                }
+            }
+            indexOs = new OutputStreamWriter(new FileOutputStream(index, true), StandardCharsets.UTF_8);
+        }
+    }
+
+    @Override public BuildListener overallListener() throws IOException, InterruptedException {
+        return new StreamBuildListener(new IndexOutputStream(null), StandardCharsets.UTF_8);
+    }
+
+    @Override public TaskListener nodeListener(FlowNode node) throws IOException, InterruptedException {
+        return new StreamTaskListener(new IndexOutputStream(node.getId()), StandardCharsets.UTF_8);
+    }
+
+    private void checkId(String id) throws IOException {
+        assert Thread.holdsLock(this);
+        if (!Objects.equals(id, lastId)) {
+            long pos = os.getChannel().position();
+            if (id == null) {
+                indexOs.write(pos + "\n");
+            } else {
+                indexOs.write(pos + " " + id + "\n");
+            }
+            indexOs.flush();
+            lastId = id;
+        }
+    }
+
+    private final class IndexOutputStream extends OutputStream {
+
+        private final String id;
+
+        IndexOutputStream(String id) throws IOException {
+            this.id = id;
+            open();
+        }
+
+        @Override public void write(int b) throws IOException {
+            synchronized (FileLogStorage.this) {
+                checkId(id);
+                os.write(b);
+            }
+        }
+
+        @Override public void write(byte[] b) throws IOException {
+            synchronized (FileLogStorage.this) {
+                checkId(id);
+                os.write(b);
+            }
+        }
+
+        @Override public void write(byte[] b, int off, int len) throws IOException {
+            synchronized (FileLogStorage.this) {
+                checkId(id);
+                os.write(b, off, len);
+            }
+        }
+
+        @Override public void flush() throws IOException {
+            os.flush();
+        }
+
+        @Override public void close() throws IOException {
+            if (id == null) {
+                openStorages.remove(log);
+                try {
+                    os.close();
+                } finally {
+                    indexOs.close();
+                }
+            }
+        }
+
+    }
+
+    @Override public AnnotatedLargeText<FlowExecutionOwner.Executable> overallLog(FlowExecutionOwner.Executable build, boolean complete) {
+        return new AnnotatedLargeText<FlowExecutionOwner.Executable>(log, StandardCharsets.UTF_8, complete, build) {
+            @Override public long writeHtmlTo(long start, Writer w) throws IOException {
+                try (BufferedReader indexBR = index.isFile() ? Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8) : new BufferedReader(new NullReader(0))) {
```

Comment:
Seems that `new NullReader(0)` is buggy.

Reply:
…or not exactly. Apparently it throws an `IOException` (not `EOFException`) if you call `read()` after having already gotten -1 ~ EOF, unlike every other `Reader` implementation I can find which just continues to return -1. In other words, it is stricter, which is fine since it turned up a mistake in the caller here.